casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala

```diff
@@ -20,54 +22,72 @@ object ProtoUtil {
    * c is in the blockchain of b iff c == b or c is in the blockchain of the main parent of b
    */
   // TODO: Move into BlockDAG and remove corresponding param once that is moved over from simulator
-  @tailrec
-  def isInMainChain(internalMap: Map[BlockHash, BlockMessage],
-                    candidate: BlockMessage,
-                    target: BlockMessage): Boolean =
+  def isInMainChain[F[_]: Monad: BlockStore](candidate: BlockMessage,
+                                             target: BlockMessage): F[Boolean] =
     if (candidate == target) {
-      true
+      true.pure[F]
     } else {
-      (for {
-        hdr        <- target.header
-        parentHash <- hdr.parentsHashList.headOption
-        mainParent <- internalMap.get(parentHash)
-      } yield mainParent) match {
-        case Some(parent) => isInMainChain(internalMap, candidate, parent)
-        case None         => false
+      val maybeMainParentHash = ProtoUtil.parentHashes(target).headOption
+      maybeMainParentHash match {
+        case Some(mainParentHash) =>
+          for {
+            mainParent <- BlockStore[F].get(mainParentHash)
+            isInMainChain <- mainParent match {
+                              case Some(parent) => isInMainChain[F](candidate, parent)
+                              case None         => false.pure[F]
+                            }
+          } yield isInMainChain
+        case None => false.pure[F]
       }
     }
 
-  @tailrec
-  def getMainChain(internalMap: Map[BlockHash, BlockMessage],
-                   estimate: BlockMessage,
-                   acc: IndexedSeq[BlockMessage]): IndexedSeq[BlockMessage] = {
-    val parentsHashes       = ProtoUtil.parents(estimate)
+  def getMainChain[F[_]: Monad: BlockStore](
+      estimate: BlockMessage,
+      acc: IndexedSeq[BlockMessage]): F[IndexedSeq[BlockMessage]] = {
+    val parentsHashes       = ProtoUtil.parentHashes(estimate)
     val maybeMainParentHash = parentsHashes.headOption
-    maybeMainParentHash flatMap internalMap.get match {
-      case Some(newEstimate) =>
-        getMainChain(internalMap, newEstimate, acc :+ estimate)
-      case None => acc :+ estimate
-    }
+    for {
+      mainChain <- maybeMainParentHash match {
+                    case Some(mainParentHash) =>
+                      for {
+                        updatedEstimate <- unsafeGetBlock[F](mainParentHash)
+                        mainChain       <- getMainChain[F](updatedEstimate, acc :+ estimate)
+                      } yield mainChain
+                    case None => (acc :+ estimate).pure[F]
+                  }
+    } yield mainChain
   }
 
-  @tailrec
-  def findJustificationParentWithSeqNum(b: BlockMessage,
-                                        internalMap: Map[BlockHash, BlockMessage],
-                                        seqNum: SequenceNumber): Option[BlockMessage] =
+  def unsafeGetBlock[F[_]: Monad: BlockStore](hash: BlockHash): F[BlockMessage] =
```

Comment:
Do we really need this function? It's a little annoying to have to handle the `Option` at the callsite, but maybe worth it? At the very least it seems like we could use `MonadError` here instead of throwing. Also it's probably better to raise and `Exception` as opposed to an `Error`. Per oracle java docs: "An Error is a subclass of Throwable that indicates serious problems that a reasonable application should not try to catch" vs "The class Exception and its subclasses are a form of Throwable that indicates conditions that a reasonable application might want to catch."

Reply:
For almost all (if not all) calls to unsafeGetBlock, the only reason you would get the error is because you've decided to manually go into the LMDB store and delete some blocks for whatever reason. In any case, I've changed it to an Exception.